### ISSUE TITLE
Enum values changing

### ISSUE THREAD

**Issue by AlexWayfer** [7even/db_schema#5]
Hello!

I like your idea and I want to help.

> Reordering and deleting values from enum types is not supported.

You can do it!

Sequel migration example:

```ruby
Sequel.migration do
  no_transaction

  up do
    add_enum_value(:enum_locale, 'fr')
  end

  down do
    enum_values = schema(:users).to_h[:locale][:enum_values]

    alter_table :users do
      set_column_type :locale, String
    end

    self[:users].where(locale: 'fr').update(locale: 'en')

    drop_enum :enum_locale
    create_enum :enum_locale, enum_values - %w(fr)

    alter_table table do
      set_column_type(
        :locale,
        :enum_locale,
        using: 'locale::enum_locale'
      )
    end
  end
end
```

**Comment by 7even** [comment]
Hi.

I am aware of this approach with temporary converting the fields to varchar and recreating the enum with a new set of values. I've been considering to use it in DbSchema but the problem is, schema definition does not declare how to convert existing field values (`self[:users].where(locale: 'fr').update(locale: 'en')` in your example) to new enum.

Running the migration without a transaction also doesn't seem like a good idea. Postgres doesn't allow adding values to enums in a transaction so I had to [move these operations outside](https://github.com/7even/db_schema/blob/master/lib/db_schema/runner.rb#L36-L38); not only this may cause problems in the worst case (if it fails, other operations won't be rolled back) but it also doesn't allow you to add a new value to existing enum and use it in a default value of some field - all in one migration. Currently you have to make this in 2 steps - first add a new value, then apply the schema, then use the new value as a default value in a field, them apply the schema again.

I think we can implement a "temporary varchar" solution and use it for all changes of enum values - both adding and removing values (reordering will also be covered by the way). It's somewhat hacky but it solves several problems. Converting values in fields, on the other hand, falls out of schema definition approach - this will be possible with conditional migrations (see slide 22 of [presentation](https://www.slideshare.net/VsevolodRomashov/database-schema-management-in-ruby-apps)) once I implement them in DbSchema.

**Comment by AlexWayfer** [comment]
> see slide 22 of presentation

I was on this presentation :)

> this will be possible with conditional migrations

Yes, I was thinking about them.

> Running the migration without a transaction also doesn't seem like a good idea.

This is required only for adding values, not for `down` migration (in which there is a removal of the value, about which I say).

**Comment by 7even** [comment]
> This is required only for adding values, not for down migration (in which there is a removal of the value, about which I say).

Yes, I understand. I'm just saying we can use the varchar approach not only to remove values but also to add them - so that everything runs in a transaction.

**Comment by AlexWayfer** [comment]
Oh, thank you, now I understand.

Yes, you can. This will be a more cumbersome code, but probably more secure.

Cumbersome is of little importance when working inside an external library and a one-time repetition :)

**Comment by 7even** [comment]
@AlexWayfer released in 0.2.5. Took a lot longer than I expected.

**Comment by AlexWayfer** [comment]
@7even Good work! Congratulations on your success :)
